### PR TITLE
Speedup query to do ApiKey grants

### PR DESCRIPTION
**Pull request by amiedes**
Related: https://app.clubhouse.io/cartoteam/story/117080/performance-issues-with-datasets-r-scope

**What does this PR do?**

Previously, when generating the `GRANT`s associated to an API key, we were doing a query similar to the following to retrieve the sequences associated with each of the tables the API key had access to:

```sql
SELECT
  n.nspname, c.relname
FROM
  pg_depend d
  JOIN pg_class c ON d.objid = c.oid
  JOIN pg_namespace n ON c.relnamespace = n.oid
WHERE
  d.refobjsubid > 0 AND
  d.classid = 'pg_class'::regclass AND
  c.relkind = 'S'::"char" AND
  d.refobjid IN (
    (QUOTE_IDENT('amiedes') || '.' || QUOTE_IDENT('africa_adm0'))::regclass
  );
```

For an API key with access to 150 tables, this query took ~200ms. This means for the 150 tables it was around 35 seconds.

This PR combines all the queries to retrieve the list of sequences to give the `GRANT`s in a single query, so it is faster:

```sql
SELECT
  n.nspname, c.relname
FROM
  pg_depend d
  JOIN pg_class c ON d.objid = c.oid
  JOIN pg_namespace n ON c.relnamespace = n.oid
WHERE
  d.refobjsubid > 0 AND
  d.classid = 'pg_class'::regclass AND
  c.relkind = 'S'::"char" AND
  d.refobjid IN (
    (QUOTE_IDENT('amiedes') || '.' || QUOTE_IDENT('africa_adm0'))::regclass,
    (QUOTE_IDENT('amiedes') || '.' || QUOTE_IDENT('air_quality_data_2017'))::regclass,
    (QUOTE_IDENT('amiedes') || '.' || QUOTE_IDENT('air_quality_stations_2017'))::regclass,lass
  );
```

Now the same query takes ~2 seconds.

### CODE REVIEW

File: app/models/carto/api_key.rb [hunk 1]
```diff
@@ -371,15 +373,19 @@ def grant_ownership_role_privileges
...
         Carto::TableAndFriends.apply(db_connection, tp.schema, tp.name) do |schema, table_name, qualified_name|
           db_run("GRANT #{tp.permissions.join(', ')} ON TABLE #{qualified_name} TO \"#{db_role}\"")
-          sequences_for_table(schema, table_name).each do |seq|
-            db_run("GRANT USAGE, SELECT ON SEQUENCE #{seq} TO \"#{db_role}\"")
-          end
+          tables << { schema: schema, table_name: table_name }
         end
       end
 
+      user.db_service.sequences_for_tables(tables).each do |sequence|
```
Comment (amiedes) [inline app/models/carto/api_key.rb:385]: The old `sequences_for_table(schema, table_name)` did N queries to retrieve all the sequences for all the tables. The new `user.db_service.sequences_for_tables(tables)` does the same but in a single query.

File: app/models/carto/api_key.rb [hunk 2]
```diff
@@ -585,26 +591,6 @@ def remove_from_redis(key = redis_key)
       redis_client.del(key)
     end
 
-    def sequences_for_table(schema, table)
```
Comment (amiedes) [inline app/models/carto/api_key.rb:588]: The new version of this (`sequences_for_tables`) has been moved to the DB service.

File: spec/models/carto/user_db_service_spec.rb [hunk 3]
```diff
@@ -1,36 +1,62 @@
...
-  let(:user) { create(:carto_user) }
-  let(:db_service) { Carto::UserDBService.new(user) }
+  include Carto::Factories::Visualizations
+
+  let(:user) { create(:valid_user).carto_user }
+  let(:db_service) { user.db_service }
 
   describe '#public_user_roles' do
-    subject { db_service.public_user_roles }
+    subject(:public_user_roles) { db_service.public_user_roles }
```
Comment (amiedes) [inline spec/models/carto/user_db_service_spec.rb:10]: These are mostly linter fixes, the important part is under `describe '#sequences_for_tables' do`

**Review by thedae — APPROVED** [review]
LGTM :+1:

**Comment by amiedes** [comment]
**Acceptance in staging** ✅ 

API key creation continues to work as expected:

```json
{
    "id": "75f6c827-6933-4b4f-bf9f-78e29aad4a2e",
    "token": "JWTDtOqbc_d3IogeGIY4wQ",
    "user_id": "8c8ec7bb-29a8-432c-90ee-5d2a1ae472de",
    "type": "regular",
    "name": "Test",
    "db_role": "carto_role_94bb1a0408a21405ba67d88fbb21019b",
    "db_password": "2063f7789c301dc4cc981c560ce0870a992eafdd",
    "grants": [
      {
        "type": "apis",
        "apis": [
          "maps",
          "sql"
        ]
      },
      {
        "type": "database",
        "tables": [
          {
            "name": "dataset_ciudades_2",
            "schema": "amiedes-sequences1",
            "permissions": [
              "select",
              "update",
              "insert",
              "delete"
            ]
          },
          {
            "name": "dataset_ciudades",
            "schema": "amiedes-sequences1",
            "permissions": [
              "select",
              "update",
              "insert",
              "delete"
            ]
          },
          {
            "name": "dataset_ciudades_1",
            "schema": "amiedes-sequences1",
            "permissions": [
              "select",
              "update",
              "insert",
              "delete"
            ]
          }
        ],
        "schemas": [

        ]
      }
    ],
    "created_at": "2020-12-09 09:53:51 UTC",
    "updated_at": "2020-12-09 09:53:51 UTC"
  }
```

Will do more performance testing in production with the real application.

**Pull request merged**